### ISSUE TITLE
Java Ink

### ISSUE THREAD

**Issue by micabytes** [bladecoder/blade-ink#1]
Cool to have someone else interested in using Ink in Java. There doesn't seem to be many of us.

I think it is rather unnecessary to have multiple different Java ports of Ink, so ideally, I would like to see the code bases of jink and blade-ink merged so that they end up as just as single library.

Although I'm not a fan of ink's JSON format, the ink architecture is built centrally around it. That could become a problem if it changes too often, though I hope that this will not happen. Assuming inkle is careful about that, your approach of a straight "runtime" port should hopefully prove the better strategy, and I'm willing to abandon most of my code and merge towards your code base. 

I'm already using jInk in several apps, though, so before I can merge into the codebase, there are a couple of issues that I need to figure out/look into.
### 1. Java objects

Basically a must-have requirement for me. For a game that uses lots of in-game state (i.e., every game that I'm interested in doing), the `EXTERN FUNCTION` syntax creates way too much boilerplate script and java code.

jInk permits the usage of Java objects as variables and allows method calls on Java methods (implemented very simply using reflection). This allows me to insert a "world" object at the start of a script, extract objects as required through that object, and get/change the world state as required by direct method calls.

Need to check if dots are allowed in variables and function names (i.e., is `world.getPlayer()` a valid function name). If so, it would probably be doable without too huge changes. The ink -> json would just consider objects and methods calls to be regular variables and functions, and all the magic should be possible to do in the runtime in the variable/function evaluation code. If not, then I'm actually not sure how to implement this.
### 2. Sequence on variable/value

Discussed https://github.com/inkle/ink/issues/57

I use this heavily, and implemented it pretty early in jInk. Would need to have inkle's acceptance of this into the ink language, obviously, but if they accept it, is trivial to implement.
### 3. Special Instructions

Because jInk runs on ink script files, I've used comments to implement some special functionality into the engine - e.g., to show images, etc. 

This will of course not work through Json, as the comments get stripped away. But most of this functionality could probably be moved to methods (especially with 1. above) or a be replaced by the recently proposed tag system: https://github.com/inkle/ink/issues/122

I do have some behavior tree-like behavior that I've implemented using Ink scripts, that might be difficult to do with tags but it should be possible to find some other way to do this functionality.
### 4. Json Decoding Footprint

Just based off a cursory glance at the Json decoding, it seems like you are not using a streaming parser for the Json decoding. If so, I would be a bit worried about the decoding footprint on Android devices - I do all of my Java game coding for Android, and I've occasionally found decoding of large Json files to be a major issue if they're not streamed, since there are usually other things also requiring lots of memory (images, sound, etc). OOM errors are a pain.
### 5. Handling changes

One issue I haven't seen discussed yet on Ink's forums is how to handle changes in the game code. Continuous iteration or debugging on game code may often result in game scripts changing after release. This requires some robust save/load functionality to prevent crashes when loading story state from a save file since the knots, stitches and choices saved may no longer exists.

Something to look into - I have issues enough with it today when converting directly from Ink scripts; I can't imagine the problem is much easier in Json.
### 6. QA/Testing

Always important, of course. I'd want to extend the test set (merging the two test sets would be a good start), and hooking things up to Travis CI or similar should be a must. Ideally combined with mutual code reviews of changes, given how critical the code is to the stability of any game built on top of it.

Your thoughts? I wil probably think of more issues later, but these are the ones that come to mind right away. 1. and 3. are the two points I find a bit worrying; the latter issues are mostly about working out a strategy for a common library, which I think we'll have to do, if this is to make sense for me. Not much point for me to do the work to migrate code to blade-ink's interface if we're going to end up with two forks of the codebase in a year from now.

**Comment by bladecoder** [comment]
Hi,

The **blade-ink** is a class by class Java port of the C# runtime engine. This approach makes easy the synchronization of new features with the reference implementation.

Right now, you have a parser with runtime capabilities. We could adapt your parser to use the **blade-ink** and generate `json`. The equivalent to the official `inklecate`. With this, we will have a complete implementation (compiler + rt engine) similar to the official C# implementation. What do you think?

Here you have my contribution to your issues:

### 1. Java objects

This can be very cool and can be added to the engine in a similar way to external functions. I see some possible problems with this feature:

  + External functions have a feature that allows use fallback functions inside the ink file if the function does not exists. This allows testing the story inside `inklecate` or the new `inky`. If we use external objects, testing the story outside the game will not be possible.

  + The source code will not compile with the official `inklecate`.

  + The .json will not run with the C# or the JS runtimes.

### 2. Sequence on variable/value

The runtime does not have to be modified to support this feature, this falls in the compiler side. As in the previous issue, the source code will not compile with the official `inklecate` but the generated .json will be compatible.

The better approach would be that the official Ink supports this feature.

### 3. Special Instructions

I think that in order to address this, we need to push the official Ink to adopt the solution proposed in https://github.com/inkle/ink/issues/122

### 4. Json Decoding Footprint

I use the same code used in the C# engine. I don't know if this will be an issue. I plan to do an Android player in the next weeks, we can test a very large file in a real device. 


### 5. Handling changes

In this case, I think that the .json format is an advantage. Saved games store variables, counters, callstack and the current container. All this 'should' work well when adding or deleting content.

### 6. QA/Testing

You have a good testing set. I started to add your tests to my runtime and I want to add all C# tests and 'inkjs' too. I agree with you about have a good set of tests. Maybe we can push to have a reference test set to share with all implementations (C#, Java and JS).

As you can see, I would like that all implementations work with @Inkle together to share efforts and push the standard instead of branch it. What do you think?

**Comment by micabytes** [comment]
Awesome; I think we are much of the same mind - agree that it would be ideal to maintain compatibility with standard ink. Not sure my parser would be all that reusable as Json generator since I use a different internal model than the Inkle guys do, but could be worth looking into at some point.

1. My thought was that one could allow fallback functions for this functionality in the same way as for regular external functions. That would have required dot to be valid part of a function/variable name, though, and unfortunately - based on a quick check - that is not the case. Absent that syntax, I'm not really sure how to deal with this issue. Probably need to discuss this with Jon and Joe.

2. Agreed.

3. Agreed.

4. Ah. It may not be a big problem for Ink's Json, though sometimes every megabyte counts. I use Jackson's streaming API for pretty much everything in Android myself now.

5. Not sure that's possible - containers for which counters exist may be deleted, even the current container. At the moment, the only solution to "handling" iterative development of a story without side-effects that I can see would be to save the entire story in the savegame (such that latter changes do not affect existing savegames). Or perhaps I'm overlooking something.

6. If you're not averse to it, I would suggest just adding the Specs2/Scala framework wholesale doing a copy-replace on the interface. Scala is interoperable with Java, the Scala code used in this context is trivial, and Specs2 is pretty awesome (at my work, test coverage shot up hugely once we introduced it - it _almost_ becomes enjoyable to write tests).

Issue 1 seems like it may be the biggest problem initially. I'll try to put together some thoughts on that, and open an issue on the main inkle repo to hear their thoughts.

**Comment by micabytes** [comment]
Hi @bladecoder , apologies for not following up on this sooner (been busy, etc).

Had some discussion with the Inkle guys some weeks ago, wrt the issues mentioned above. What I got from that conversation was that object functionality will not be coming soon, and they're unwilling to change the syntax (i.e., make dot a valid part of the function notation) until they know how this would be done. The new sequence mechanic could be added, and tags will come at some point.

Unfortunately, #1 is the make-or-break feature as far as I am concerned. I don't see any good way to implement this without breaking with the ink format and the alternatives I've discussed with them are simply not useful from my point of view - too much additional boilerplate added in the script, and too much additional game engine coding around the Ink script.

Simply put, I would have to rewrite all my existing ink scripts (adding more boilerplater, making them significantly more unreadable) and rebuild/extend significant parts of my game engine to get this to work, in the hope that there will be some long-term benefits down the line. Much as I love the idea of collaborating around an ink-based open-source project, I can't justify that cost at this time when my existing implementation does what I need, and can be easily extended to cover the foreseeable features I require. I built a game engine to write games, so the latter is what I want to focus my time on.

So that's my conclusion at the current time. Perhaps things may change in future.

**Comment by bladecoder** [comment]
I'm going to close this. Each engine is following its own path.
